tests/10apidoc/13ui-auth.pl

```diff
@@ -29,17 +82,132 @@ sub wait_for_cas_request
    });
 }
 
+# Log into Synapse via CAS.
+#
+# This takes the following parameters:
+#  * The Matrix user ID to log in with.
+#  * The CAS ticket to use. This can be any string.
+#  * The HTTP object.
+#  * The Homeserver Info object.
+#  * The XML CAS response. See generate_cas_response.
+#  * Any additional parameters will be passed as part of the login request.
+sub matrix_login_with_cas
+{
+   my ( $user_id, $http, $homeserver_info, $cas_response, %params ) = @_;
+
+   # the redirectUrl we send to /login/sso/redirect, which is where we
+   # hope to get redirected back to
+   my $REDIRECT_URL = "https://client?p=http%3A%2F%2Fserver";
+
+   my $HS_URI = $homeserver_info->client_location . "/_matrix/client/r0/login/cas/ticket?redirectUrl=" . uri_escape($REDIRECT_URL);
+
+   # the ticket our mocked-up CAS server "generates"
+   my $CAS_TICKET = "goldenticket";
+
+   # step 1: client sends request to /login/sso/redirect
+   # step 2: synapse should redirect to the cas server.
+   Future->needs_all(
+      wait_for_cas_request( "/cas/login" ),
+      $http->do_request(
+         method => "GET",
+         uri    => "/r0/login/sso/redirect",
+         params => {
+            redirectUrl => $REDIRECT_URL,
+         },
+      ),
+   )->then( sub {
+      my ( $cas_request, $_cas_response ) = @_;
+      log_if_fail( "Initial CAS request query:",
+                   $cas_request->query_string );
```

Comment:
We tend not to put parens around `log_if_fail`. Do we need them here?

Reply:
Nope. I've made this change in the entire file.